pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+		// from the resolver.
+		if errors.IsNotFound(err) {
+			r.res.Delete(request.Name)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Get a copy of the DNSNameResolver resource.
+	dnsNameResolver = dnsNameResolver.DeepCopy()
+
+	// Check if the grace period is over for some of the IP addresses after the expiration of their respective TTLs. If so,
+	// remove those IP addresses and update the status of the resource.
+	if removalOfIPsRequired(&dnsNameResolver.Status) {
+		if err := r.client.Status().Update(ctx, dnsNameResolver); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the TTLs of some of the IP addresses have expired. If so, requeue the reconcile request after the minimum remaining
+	// time until the grace period gets over among that of the IP addresses with expired TTLs.
+	if ttlExpired, remainingTime := reconcileRequired(&dnsNameResolver.Status); ttlExpired {
+		return reconcile.Result{Requeue: true, RequeueAfter: remainingTime}, nil
+	}
+
+	dnsName := string(dnsNameResolver.Spec.Name)
+	matchesRegular := !isWildcard(dnsName)
+
+	// Add the DNS name in the spec of the DNSNameResolver resource to the resolver with empty resolved addresses. This is done
+	// to ensure that the resolver sends the resolution request for the DNS name if it is getting added for the first time.
+	r.res.Add(dnsName, []networkv1alpha1.DNSNameResolverResolvedAddress{}, matchesRegular, request.Name)
+
+	// Iterate through each of the resolved names matching the DNS name and add them to the resolver along with the current IP addresses.
+	for _, resolvedName := range dnsNameResolver.Status.ResolvedNames {
+		r.res.Add(string(resolvedName.DNSName), resolvedName.ResolvedAddresses, matchesRegular, request.Name)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// isWildcard checks if the domain name is wildcard.
+func isWildcard(dnsName string) bool {
+	trimmedString := strings.TrimPrefix(dnsName, "*")
+	return trimmedString != dnsName
+}
+
+// removalOfIPsRequired checks whether the TTL of any IP address, associated to a DNS name
+// matching the DNSNameResolver object, has expired and the grace period after the TTL
+// expiration is also over. If so, then those resolved addresses are removed from the
+// resolved name details of the DNS name.
+func removalOfIPsRequired(status *networkv1alpha1.DNSNameResolverStatus) bool {
+	updated := false
+	for index, resolvedName := range status.ResolvedNames {
+		updateResolvedAddresses := []networkv1alpha1.DNSNameResolverResolvedAddress{}
+		for _, resolvedAddress := range resolvedName.ResolvedAddresses {
+			if time.Now().Before(resolvedAddress.LastLookupTime.Time.Add(time.Second*time.Duration(resolvedAddress.TTLSeconds) + ipRemovalGracePeriod)) {
+				updateResolvedAddresses = append(updateResolvedAddresses, resolvedAddress)
+			} else {
```

Comment:
Why is this not marked updated when the list of updateResolvedAddresses grows?

Reply:
The updateResolvedAddresses holds those addresses which should still remain in the status. We are marking it updated only if at least one resolved address is not added to the updateResolvedAddresses slice. This means that those addresses will be removed from the status when the object is updated.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2